src/npcmove.cpp

```diff
@@ -1759,46 +1701,103 @@ bool npc::recharge_cbm()
+    healing_options try_to_fix;
+    try_to_fix.clear_all();
 
-    if( need_heal( *this ) ) {
-        if( !use_bionic_by_id( bio_nanobots ) ) {
-            return npc_heal;
+    for( int i = 0; i < num_hp_parts; i++ ) {
+        const hp_part part = static_cast<hp_part>( i );
+        const body_part bp_wounded = hp_to_bp( part );
+
+        if( c.has_effect( effect_bleed, bp_wounded ) ) {
+            try_to_fix.bleed = true;
+            return try_to_fix;
         }
-    } else {
-        deactivate_bionic_by_id( bio_nanobots );
-    }
 
-    if( get_perceived_pain() >= 15 ) {
-        if( !activate_bionic_by_id( bio_painkiller ) && has_painkiller() && !took_painkiller() ) {
-            return npc_use_painkiller;
+        if( c.has_effect( effect_bite, bp_wounded ) ) {
+            try_to_fix.bite = true;
+            return try_to_fix;
+
+        }
+        // NPCs don't reapply bandages
+        if( !c.has_effect( effect_bandaged, bp_wounded ) ) {
+            int part_threshold = 50;
+            if( part == hp_head ) {
+                part_threshold += 20;
+            } else if( part == hp_torso ) {
+                part_threshold += 10;
+            }
+            part_threshold = std::min( 80, part_threshold );
+            part_threshold = part_threshold * c.hp_max[i] / 100;
+
+            if( c.hp_cur[i] <= part_threshold ) {
+                try_to_fix.bandage = true;
+                return try_to_fix;
+            }
         }
-    } else {
-        deactivate_bionic_by_id( bio_nanobots );
-    }
 
-    // If there are no healing items, there's nothing to do
-    if( ai_cache.can_heal.bandage || ai_cache.can_heal.bleed || ai_cache.can_heal.bite ||
-        ai_cache.can_heal.infect ) {
-        if( is_player_ally() && need_heal( g->u ) ) {
-            ai_cache.ally = g->shared_from( g->u );
-            return npc_heal_player;
+        if( c.has_effect( effect_infected, bp_wounded ) ) {
+            try_to_fix.infect = true;
+            return try_to_fix;
         }
+    }
 
-        const std::vector<npc *> hurt_allies = g->get_npcs_if( [&]( const npc & guy ) {
-            return guy.getID() != getID() && guy.is_ally( *this ) && posz() == guy.posz() &&
-                   need_heal( guy ) && rl_dist( pos(), guy.pos() ) <= SEEX * 2 && sees( guy.pos() );
-        } );
+    return try_to_fix;
+}
 
-        for( npc *guy : hurt_allies ) {
-            ai_cache.ally = g->shared_from( *guy );
-            return npc_heal_player;
+npc_action npc::address_needs( float danger )
+{
+    // rng because NPCs are not meant to be hypervigilant hawks that notice everything
+    // and swing into action with alarming alacrity.
+    // no sometimes they are just looking the other way, sometimes they hestitate.
+    // ( also we can get huge performance boosts )
+    if( one_in( 3 ) && get_skill_level( skill_firstaid ) > 0 ) {
+        healing_options try_to_fix_me = patient_assessment( *this );
```

Comment:
Let's break self-healing out from this loop.  NPCs always assess themselves, even with first aid 0, and let their nanobots run until they aren't injured.